docs/proposal-001_add-ons.md

```diff
@@ -0,0 +1,163 @@
+# Desing Proposal #001: Add-ons
+
+## What's an add-ons?
+
+An add-on extends functionality of a Kubernetes cluster, it may consist of a workload and/or configuration within the give cluster or the cloud provider. The workload, if present, would may classify as an operator or a controller, however that is not neccessary.
+
+The difference between add-ons and anything a user decided to run themselves, is that from user's point of view, add-on is something they don't need to maintain directly. A cluster provider or bootstrap tool (i.e. eksctl), is expected to ensure ease of use, by providing minumum viable (that configuration to ensure end-to-end integration and compatibility), as well as catering for reconfiguration and upgrades.
+
+## Design Requirements and Goals
+
+Add-on must expose all of the following attributes:
+
+- name (fixed) *[TBD]*
+- version tag or revision (has default value that user can override)
+- source URL (has default value that user can override)
+- set of customisations (no default value, user defined)
+
+Internally, an add-on may consist of any and/or all of the following elements:
+
+- Kubernetes workload definitions
+- Kubernetes configuration objects - PersistentVolumes, ConfigMaps, ...etc
+- cloud provider resource definitons (e.g. CloudFormation sub-resources)
+- dependencies (supported Kubernetes versions and corresponding cloud providers, list of references to other add-ons)
+- incompatibilities (supported Kubernetes versions and corresponding cloud providers, list of reference to other add-ons)
+
+In other words, it's ought to be possible to define an add-on that only includes:
+
+- dependencies on other add-ons
+- cloud provider resource definitions
+- Kubernetes configuration objects
+- Kubernetes workload definitions
+
+The implementation must also ensure:
+
+- fully deterministic behaviour for a given version of eksctl, thereby it should have
+    - strict versioning of built-in add-ons by default
+    - any externally-sourced config has to be pinned down or vendored *[TBD]*
+    - there should be an option to use latest version or altenative source URL
+- user should be able:
+    - remove any add-on (including it's dependencies) at any time
+    - customise an add-on in an arbitrary fashion
+    - upgrade an add-on
```

Comment:
This is going to be hard. In-place, no downtime, no data-loss upgrade in the general case is very hard.

Upgrade to what versions? I was assuming each eksctl would come with one vendored copy of each addon's source (e.g. helm chart), tested with that version of EKS and the supporting TF. Would upgrading the addon require upgrading EKS et al too? Is that possible atm?

Reply:
I agree this requirement is fairly difficult to concur and will be an ongoing battle, but I believe the onus of upgradeability should be on the add-on itself being either backwards compatible or document breaking version changes.

Just following semantic versioning could allow us to warn a user of potentially breaking changes.

Reply:
> I was assuming each eksctl would come with one vendored copy of each addon's source

Yes, that's the idea. But you'd be able to upgrade to some other version from external (read: untested) source.

> Would upgrading the addon require upgrading EKS et al too?

Only if it requires newer version of Kubernetes, but even that I don't think it'd make sense to upgrade automatically.